rpi/src/drivers/fat32.rs

```diff
@@ -149,92 +326,277 @@ impl Fat32{
+                let mut segment = self.fat_table_cache.as_slice().into_iter().find(|f|f.start_index == existing_entry.get_first_cluster_index()).unwrap().clone();
+                match segment.state {
+                    FatSegmentState::Allocated => segment.len += 1,
+                    FatSegmentState::AllocatedEof => {},
+                    _ => core::panic!("Error recevied not allocated segment"),
+                }
+                let mut fat_buffer:FatBuffer = FatBuffer::new(self.fat_info, existing_entry.get_first_cluster_index() as usize, Some(segment.len as usize), &mut self.disk);
+                let mut current_cluster = existing_entry.get_first_cluster_index();
+                let mut cluster_count = 0;
+                while cluster_count < segment.len{
+                    let start_sector = self.get_cluster_start_sector_index(current_cluster);
+                    let start_index = (sectors_per_cluster * cluster_count) as usize * SECTOR_SIZE;
+                    let end_index = start_index + (sectors_per_cluster * SECTOR_SIZE as u32) as usize;
+                    let _ = self.disk.write(start_sector, &mut content[start_index..end_index]);
+
+                    current_cluster = fat_buffer.read().ok().unwrap();
+                    cluster_count += 1;
+                }
                 return;
             }
-            current_cluster = fat_entry;
-            cluster_counter += 1;
-            fat_entry_offset += FAT_ENTRY_SIZE;
-            if fat_entry_offset >= SECTOR_SIZE as usize{
-                fat_entry_offset = 0;
-                fat_sector_number += 1;
-                let _ = self.disk.read(fat_sector_number, &mut fat_buffer);
+        }
+
+        // create a new file by allocating place in the root dir and then picking some free fat segment to use it's clusters
+        let new_dir_entry = match self.root_dir_cache.as_mut_slice().into_iter().find(|d|d.file_name[0] == DELETED_DIR_ENTRY_PREFIX){
+            Some(dir) => {
+                dir.file_name = name;
+                dir.file_extension = extension;
+                dir
             }
+            None => {
+                // Check the root dir allocation size to check it needs to be reallocated
+                let root_dir_entry = self.root_dir_cache.as_slice().into_iter().find(|d|d.attributes == ATTR_VOLUME_ID).unwrap();
+                if root_dir_entry.size as usize >= self.root_dir_cache.len() * size_of::<FatShortDirEntry>() {
+                    core::panic!("driver do not support resizing of the root dir");
+                }
+                // Allocate new entry in the root dir
+                self.root_dir_cache.insert(self.root_dir_cache.len() - 1, FatShortDirEntry::new(name, extension, content.len() as u32));    // write to the last place pusing the last item to index len(), in order to keep the DIR_EOF last
+                self.root_dir_cache.last_mut().unwrap()
```

Comment:
This is a bug, this returns the last item but the last item is not the newly created entry but rather the last entry (which is the dir eof mark).